renewCerts.sh

```diff
@@ -0,0 +1,73 @@
+#!/bin/bash
+
+# A script to renew certificates with certbot and send chat message with result
+
+# Example usage, expected within crontab:
+# 0 18 * * SUN /home/deploy/renewCerts.sh
+
+# Requirements:
+# certbot (installed on the host, not in a container),
+# nginx already configured to point to letsencrypt certificates, and
+# Environment variables in /home/deploy/.env:
+# - REVERSE_PROXY_CONTAINER_USER
+# - WEB_SERVER_HOSTNAME
+# - AUTH_SERVER_HOSTNAME
+# - ZULIP_BASE_URL
+# - ZULIP_BOT_EMAIL_ADDRESS
+# - ZULIP_BOT_API_KEY
+# - ZULIP_STREAM
+# - ZULIP_TOPIC
+
+set -eo pipefail
+
+test -x "$(which certbot)"
+. /home/deploy/.env
+test ! -z "$WEB_SERVER_HOSTNAME"
+test ! -z "$REVERSE_PROXY_CONTAINER_USER"
+test ! -z "$AUTH_SERVER_HOSTNAME"
+test ! -z "$ZULIP_BASE_URL"
+test ! -z "$ZULIP_BOT_EMAIL_ADDRESS"
+test ! -z "$ZULIP_BOT_API_KEY"
+test ! -z "$ZULIP_STREAM"
+test ! -z "$ZULIP_TOPIC"
+
+set +eo pipefail
+
+function fin() {
+    # Exit. Also notify of the result via chat if an API key is present.
+    exit_code=0
+    error_message=$1
+    message="✅ Certificate renewals (or checks) for https://${AUTH_SERVER_HOSTNAME} and https://${WEB_SERVER_HOSTNAME} succeeded."
+
+    if test ! -z "$error_message"; then
+        exit_code=1
+        message="❌ Certificate renewals (or checks) for https://${AUTH_SERVER_HOSTNAME} or https://${WEB_SERVER_HOSTNAME} FAILED: ${error_message}"
+    fi
+
+    if test ! -z "$ZULIP_BOT_API_KEY"; then
+	curl -X POST ${ZULIP_BASE_URL}/api/v1/messages \
+	     -u ${ZULIP_BOT_EMAIL_ADDRESS}:${ZULIP_BOT_API_KEY} \
+	     --data-urlencode type=stream \
+	     --data-urlencode to=${ZULIP_STREAM} \
+	     --data-urlencode topic=${ZULIP_TOPIC} \
+	     --data-urlencode "content=${message}"
+    fi
+
+    exit $exit_code
+}
+
+certbot certonly -n --domain=$AUTH_SERVER_HOSTNAME --standalone --keep-until-expiring \
```

Comment:
How does the `--standalone` option ("Run a standalone webserver for authentication") work with the existing port mapping? I would have expected that to fail!

Reply:
In order for this to work, we do not listen on port 80 on the reverse proxy. Certbot does its hostname validation via port 80. After all certbot is creating a certificate for use on port 443 so to require 443 would cause deadlock. Not listening on port 80 means we unfortunately can't redirect plain http traffic on port 80 to https on 443. Then again it is 2023 so we've had a decade or more to start defaulting to `https` when filling out web addresses.

Reply:
Ah. That is a limitation, but at least I did understand all the moving parts correctly!